### PR TITLE
Replace RuntimeTarget with RuntimeHolder

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
Original Draft replace RuntimeIdentity with a new entity RuntimeTarget.
And all usage of RuntimeIdentity is replaced with RuntimeTarget. And there are a lot of changes around that.

This PR has an alternative proposal:
Instead of 
> RuntimeTarget#getNamespace returns a namespace where runtime is run or null if default one should be used
we have
> RuntimeIdentity#getNamespace returns a namespace where runtime is run. MUST NOT be null

and when Workspace API does not know which namespace workspace should use, it asks Infrastructure to evaluate it with 
```java
public class RuntimeHolder {
  private final String workspaceId;
  private final String ownerId;
  private final String ownerName;
}
```

Instead of reviewing changes in this PR it may be easier to take a look 
https://github.com/eclipse/che/compare/master...sleshchenko:extendRuntimeIdentity?expand=1
against 
https://github.com/eclipse/che/pull/15040/files

### CODE REVIEW

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/RuntimeIdentityImpl.java [hunk 1]
```diff
@@ -13,22 +13,55 @@
...
+      String ownerName,
+      String infrastructureNamespace) {
+    this.workspaceId = workspaceId;
+    this.envName = envName;
+    this.ownerId = ownerId;
+    this.ownerName = ownerName;
+    this.infrastructureNamespace = infrastructureNamespace;
+  }
+
+  public RuntimeIdentityImpl(String id, Subject subject, String envName) {
```
Comment (metlos) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/RuntimeIdentityImpl.java]: the infrastructure namespace must not be null according to javadoc, yet the ctors allow for it...

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/RuntimeHolder.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * Holds meta info about to what runtime belongs to: workspace and user info.
+ *
+ * <p>Must be used during evaluating a default namespace name for runtime.
+ *
+ * <p>TODO Proposals about better name is welcome, alternatives RuntimeHolder
+ *
+ * @author Sergii Leshchenko
+ */
+public class RuntimeHolder {
```
Comment (metlos) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/RuntimeHolder.java]: This is essentially equal to `KubernetesNamespaceFactory#PlaceholderResolutionContext`. Why not make that class top level instead?

**Review by metlos — COMMENTED** [review]
While I like the fact that this approach limits the amount of change and "looks" nicer, I think the previous `RuntimeTarget` made a better job of limiting the scope of potential breakage. The new `RuntimeIdentity#getInfrastrctureNamespace` is assumed to be non-null yet, you don't account for that in the `KubernetesRuntimeState` JPA entity which will therefore produce instances of `RuntimeIdentity` that break the invariants.

To fix that, we'd have to run a manual upgrade step during the server startup that would check for null namespaces in the backing table and fix them using the default value. Not sure if that is nicer than having an explicit divide between `RuntimeIdentity` and `RuntimeTarget`.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/RuntimeIdentityImpl.java (hunk 1 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/RuntimeIdentityImpl.java]: good catch! let me take a look at that and try to propose some solution

**Comment by sleshchenko** [comment]
> To fix that, we'd have to run a manual upgrade step during the server startup that would check for null namespaces in the backing table and fix them using the default value. Not sure if that is nicer than having an explicit divide between RuntimeIdentity and RuntimeTarget.

Good catch! a namespace is already in KubernetesRuntimeState, no migration is needed for running namespaces https://github.com/metlos/che/pull/1/commits/e8f22ebf9ad867a888b3db2d69c7d7f1053046f4#diff-a854cd84238db0a96f8e6225f0c374acR90

But there is another issue with ownerName. But I believe we are able to revert this change, and client may retrieve user name with user id. It's an issue not really depend on the namespace selection issue.

**Comment by metlos** [comment]
> Good catch! a namespace is already in KubernetesRuntimeState, no migration is needed for running namespaces e8f22eb#diff-a854cd84238db0a96f8e6225f0c374acR90

Heh, that's what I get from reviewing only by diff :smile: 

> But there is another issue with ownerName. But I believe we are able to revert this change, and client may retrieve user name with user id. It's an issue not really depend on the namespace selection issue.

Well, the issue with owner name is that it basically is a superfluous piece of information that is only needed for placeholder expansion, so IMHO it isn't really a part of the "identity". That's why I had the `RuntimeTarget` which was only used during provisioning - e.g. during the time the placeholder expansion might have been needed.

**Comment by metlos** [comment]
Btw we need to change https://github.com/eclipse/che/blob/5f1657603c03f4dff84e467a7d2c223f762de38c/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java#L382 and the associated logic in `KubernetesNamespace#prepare()` - the check for the nullity of the namespace should not be needed...

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/NamespaceResolutionContext.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    int hash = 7;
+    hash = 31 * hash + Objects.hashCode(workspaceId);
+    hash = 31 * hash + Objects.hashCode(userId);
+    hash = 31 * hash + Objects.hashCode(userName);
+    return hash;
+  }
+
+  @Override
+  public String toString() {
+    return "PlaceholderResolutionContext{"
```
Comment (metlos) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/NamespaceResolutionContext.java]: You've been caught plagiarizing :smile:

Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/NamespaceResolutionContext.java]: I did not try to hide it )
Thank IDE refactoring tools )

**Pull request merged**